### ISSUE TITLE
Support Caffe2 export/pure C(++) inference mode

### ISSUE THREAD

**Issue by vivanov879** [pytorch/pytorch#853]
Hi. Please suggest how to:
1. train a model in pytorch, save it on disk.
2. load the model in c++, call model:forward() several times without loading the model each time

**Comment by fmassa** [comment]
This thread might be relevant to you
https://discuss.pytorch.org/t/serving-model-trained-with-pytorch/168

**Comment by apaszke** [comment]
If I were to suggest something, I think I'd convert all the weights to numpy and dump them in a well established format like HDF5. These should be easily readable from C.

Our main idea is to add a model exporter, that could dump it to a Caffe2 graph, as that framework is very well optimized for production usage. And even though we don't plan to put a great amount of work on packaging production-ready models trained in PyTorch right now, there are some projects in the community that are aiming to do that, and we're going to provide them with support. You can find them in the forum link that @fmassa posted.

**Comment by taion** [comment]
Has the status or priority of this issue changed at all with the formal launch of Caffe2?

**Comment by soumith** [comment]
we're working on this, but it's still medium priority.

**Comment by ryanleary** [comment]
Any way to track progress on this or to collaborate in such a way that won't duplicate efforts @soumith?

**Comment by hasB4K** [comment]
Does that mean there is a way to get static graph information from PyTorch? Having an exporter to caffe2 would be great!

**Comment by jekbradbury** [comment]
This is in progress here https://github.com/pytorch/pytorch/pull/1815

**Comment by soumith** [comment]
fixed via http://pytorch.org/docs/master/onnx.html

**Comment by ryanleary** [comment]
@soumith is there a need for contributions either on the pytorch side or onnx side to support RNNs or is that work actively underway? Thanks!

**Comment by soumith** [comment]
cc: @zdevito @ezyang who are the ONNX Points of Contact.

**Comment by ezyang** [comment]
Hi @ryanleary! For the initial release of ONNX, we only tested that it was able to export word_language_model (https://github.com/pytorch/examples/tree/master/word_language_model), so there are almost certainly other operations to add to the exporter to make other RNNs; we're definitely accepting patches in this area. However, word_language_model has a fixed number of layers, which means that ONNX exports it in unrolled form. A bit more design work (and ONNX standardization) would be needed before we can export more dynamic RNN structures, so I would say this is lower priority for the short term.
